src/common/utils.lua

```diff
@@ -241,48 +233,152 @@ function utils.camelCase(str)
 	return str
 end
 
-utils.notices = {}
+function utils.indexOf(t, value)
+	for i, v in ipairs(t) do
+		if v == value then
+			return i
+		end
+	end
+	return -1
+end
 
-function utils.notices.credit(msg)
-	local notice = {
-		Target = msg.From,
-		Action = "Credit-Notice",
-		Recipient = msg.Recipient,
-		Quantity = tostring(1),
+--- func description
+-- @param {table} params
+-- @param {string} params.id
+-- @param {number} params.timestamp
+-- @param {string | nil} params.owner
+-- @param {table | nil} params.controllers
+function utils.register(params)
+	assert(type(params) == "table", "Register Params must be a table")
+	local id, timestamp, owner, controllers = params.id, params.timestamp, params.owner, params.controllers
+	assert(type(id) == "string", "ANT ID must be a string")
+	assert(type(timestamp) == "number", "Timestamp must be a number")
+	assert(type(owner) == "string" or owner == nil, "Owner must be a string")
+	assert(type(controllers) == "table" or controllers == nil, "Controllers must be a table")
+	ANTS[id] = {
+		Owner = owner,
+		Controllers = controllers or {},
+		RegisteredAt = timestamp,
 	}
-	for tagName, tagValue in pairs(msg) do
-		-- Tags beginning with "X-" are forwarded
-		if string.sub(tagName, 1, 2) == "X-" then
-			notice[tagName] = tagValue
+end
+
+function utils.controllerTableFromArray(t)
+	assert(type(t) == "table", "argument needs to be a table")
+	local map = {}
+	for _, v in ipairs(t) do
+		map[v] = true
+	end
+	return map
+end
+
+function utils.updateAssociations(antId, state)
+	-- Remove previous associations for old owner and controllers
+	local oldAnt = ANTS[antId]
+
+	local newOwner = state.Owner or "nilOwner"
+	local newControllers = utils.controllerTableFromArray(state.Controllers)
+
+	local newAffliates = {}
+	newAffliates[newOwner] = true
+	for user, _ in pairs(newControllers) do
+		newAffliates[user] = true
+	end
+
+	if oldAnt ~= nil then
+		local previousOwner = ANTS[antId].Owner or "nilOwner"
+		local previousControllers = ANTS[antId].Controllers
+
+		local oldAffliates = {}
```

Comment:
This could be a util function:
```
function utils.affiliatesForAnt(ant)
	local affliates = {}
	affliates[ant.Owner] = true
	for user, _ in pairs(ant.Controllers) do
		affliates[user] = true
	end
	return affiliates
end
```